vignettes/python-primer.Rmd

```diff
@@ -0,0 +1,920 @@
+my_function("Friend")
+```
+
+(Note for advanced R users: Python has no equivalent of R's argument
+"promises". Function argument default values are evaluated once, when
+the function is constructed. This can be surprising if you define a
+Python function with a mutable object as a default argument value, like
+a Python list!)
+
+```{python}
+def my_func(x = []):
+  x.append("was called")
+  print(x)
+
+my_func()
+my_func()
+my_func()
+```
+
+You can also define Python functions that take a variable number of
+arguments, similar to `...` in R. A notable difference is that R's `...`
+makes no distinction between named and unnamed arguments, but Python
+does. In Python, prefixing a single `*` captures unnamed arguments, and
+two `**` signifies that *keyword* arguments are captured.
+
+```{python}
+def my_func(*args, **kwargs):
+  print("args = ", args) # args is a tuple
+  print("kwargs = ", kwargs) # kwargs is a dictionary
+
+my_func(1, 2, 3, a = 4, b = 5, c = 6)
+```
+
+Whereas the `*` and `**` in a function definition signature *pack*
+arguments, in a function call they *unpack* arguments. Unpacking
+arguments in a function call is equivalent to using `do.call()` in R.
+
+```{python}
+def my_func(a, b, c):
+  print(a, b, c)
+
+args = (1, 2, 3)
+my_func(*args)
+
+kwargs = {"a": 1, "b": 2, "c": 3}
+my_func(**kwargs)
+
+```
+
+### Defining Classes with `class`
+
+One could argue that in R, the preeminent unit of composition for code
+is the `function`, and in Python, it's the `class`. You can be a very
+productive R user and never use R6, reference classes, or similar R
+equivalents to the object-oriented style of Python `class`'s.
+
+In Python, however, understanding the basics of how `class` objects work
+is requisite knowledge, because `class`'s are how you organize and find
+methods in Python. (In contrast to R's S3 approach, where methods are
+found by dispatching from a generic). Fortunately, the basics of
+`class`'s are accessible.
+
+Don't be intimidated if this is your first exposure to object oriented
+programming. We'll start by building up a simple Python class for
+demonstration purposes.
+
+```{python}
+class MyClass:
+  pass # `pass` means do nothing.
+
+MyClass
+type(MyClass)
+
+instance = MyClass()
+instance
+type(instance)
+```
+
+Like the `def` statement, the `class` statement binds a new callable
+symbol, `MyClass`. First note the strong naming convention, classes are
+typically `CamelCase`, and functions are typically `snake_case`. After
+defining `MyClass`, you can interact with it, and see that it has type
+`'type'`. Calling `MyClass()` creates a new object **instance** of the
+class, which has type `'MyClass'` (ignore the `__main__.` prefix for
+now). The instance prints with its memory address, which is a strong
+hint that it's common to be managing many instances of a class.
+
+In the first example, we defined an empty `class`, but when we inspect
+it we see that it already comes with a bunch of attributes (`dir()` in
```

Comment:
All methods are accessible as attributes (said another way, a method is any attribute that was a function when the `class` was constructed, since that's when all the `self` partialing magic happens).... Not sure where to work that in exactly. Perhaps a (new) subsection on attributes?

Reply:
that would be great!